### PR TITLE
[#7] [UI] As a user, I can sign in with email and password

### LINKED ISSUE #7: [UI] As a user, I can sign in with email and password
## Why

To login I would need a form to enter my credentials to use the application

## Acceptance Criteria

- Login page /form for username and validations
- Toast message must be there
- UI test
- Create a login page and form
- Validations for input forms
- Show a toast message if the user is successfully authenticated

### PR DESCRIPTION
**Pull request by Prabeshpd**
- Close #7 

## What happened 👀

- Created login feature for UI
- Created a login form 
- Used next-auth to login in the application
- Add test

## Insight 📝

Created a login form for email and password. Once the user logs in it is redirected to upload page.  Used next-auth for the login call to server actions.

> [!Note]
> The authentication is not done here for the pages as there is a separate ticket for it. 

## Proof Of Work 📹

https://github.com/nimblehq/ic-nextjs-prabesh-dahal/assets/39412015/f509732e-3d80-4175-9c17-c50353c1a196

### CODE REVIEW

File: src/validations/ui/signin.ts [hunk 1]
```diff
@@ -0,0 +1,11 @@
+import * as Yup from 'yup';
+
+/**
+ * Schema for signin.
+ */
+const signinSchema = Yup.object().shape({
```
Comment (olivierobert) [inline src/validations/ui/signin.ts:6]: Same comment as in https://github.com/nimblehq/ic-nextjs-prabesh-dahal/pull/65#discussion_r1296721496

File: src/components/Login/Form.test.tsx [hunk 2]
```diff
@@ -0,0 +1,46 @@
+import * as React from 'react';
+
+import { render, waitFor, screen } from '@testing-library/react';
+import userEvent from '@testing-library/user-event';
+
+import { Login } from './Form';
+
+describe('singup Form:', () => {
```
[the remark this answers is not in the document]
Comment (Prabeshpd) [reply src/components/Login/Form.test.tsx]: Fixed in https://github.com/nimblehq/ic-nextjs-prabesh-dahal/pull/67/commits/cb0f74cbbad6ab78df71981f74e2533c65d4ed3c

File: src/components/Login/Form.tsx [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+                    type="email"
+                    className="block w-full rounded-md border-0 py-1.5 pl-1 text-gray-900 shadow-sm ring-1 ring-inset ring-gray-300 placeholder:text-gray-400 focus:ring-2 focus:ring-inset focus:ring-indigo-600 sm:text-sm sm:leading-6"
+                    placeholder="email"
+                    name="email"
+                    required
+                    onChange={handleChange}
+                    onBlur={handleBlur}
+                  />
+                </div>
+                {errors.email && touched.email && errors.email}
```
Comment (tednguyendev) [inline src/components/Login/Form.tsx:56]: Same as [this](https://github.com/nimblehq/ic-nextjs-prabesh-dahal/pull/65#discussion_r1296667527)

File: src/components/Login/schema.ts [hunk 4]
```diff
@@ -0,0 +1,11 @@
+import * as Yup from 'yup';
+
+/**
+ * Schema for signin.
+ */
+const signinSchema = Yup.object().shape({
```
[the remark this answers is not in the document]
Comment (Prabeshpd) [reply src/components/Login/schema.ts]: Fixed in https://github.com/nimblehq/ic-nextjs-prabesh-dahal/pull/67/commits/59af91bbf1edc8e3f6dd4ecf0a670cdb7cdf80be

File: src/components/Login/Form.tsx (hunk 3 above)
Comment (Prabeshpd) [reply src/components/Login/Form.tsx]: Fixed in https://github.com/nimblehq/ic-nextjs-prabesh-dahal/pull/67/commits/59af91bbf1edc8e3f6dd4ecf0a670cdb7cdf80be

File: src/app/auth/login/page.tsx [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+  const callbackUrl = searchParams.get('callbackUrl') || '/upload';
+
+  const handleFormSubmit = async (payload: Credentials) => {
+    try {
+      await signIn('credentials', {
+        redirect: false,
+        email: payload?.email,
+        password: payload?.password,
+      });
+      toast('Logged in successfully', 'success');
```
Comment (olivierobert) [inline src/app/auth/login/page.tsx:25]: This message has not been reworked similarly to follow the earlier comments I raised 😓 Ensure messages displayed to the user are well formatted and information.

```suggestion
      toast('You have logged in successfully.', 'success');
```

File: src/app/auth/login/page.tsx [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+      router.push(callbackUrl);
+    } catch (err) {
+      toast('Provided username or password is invalid.', 'error');
```
Comment (olivierobert) [inline src/app/auth/login/page.tsx:28]:
```suggestion
      toast('The provided credentials are invalid.', 'error');
```

**Review by olivierobert — APPROVED** [review]
Just minor UX issues so pre-approving to unblock you but do the required changes

File: src/app/auth/login/page.tsx (hunk 6 above)
Comment (Prabeshpd) [reply src/app/auth/login/page.tsx]: Fixed in https://github.com/nimblehq/ic-nextjs-prabesh-dahal/pull/67/commits/a5f075f3f072945b2440d3f599a2e0ea3948f395

File: src/app/auth/login/page.tsx (hunk 5 above)
Comment (Prabeshpd) [reply src/app/auth/login/page.tsx]: Fixed in https://github.com/nimblehq/ic-nextjs-prabesh-dahal/pull/67/commits/a5f075f3f072945b2440d3f599a2e0ea3948f395

**Review by tednguyendev — APPROVED** [review]

**Pull request merged**